Build failure: odin

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

- `git clone https://github.com/NixOS/nixpkgs`
- `cd nixpkgs`
- (optionally) `git switch nixos-unstable` (happens on master and nixos-unstable)
- `nix-build -A odin`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/296369937

### Relevant log output

```console
$ nix-build -A odin
this derivation will be built:
  /nix/store/m9mg8cyy03r4lv7dfbnm0402sqrmchgp-odin-dev-2025-01.drv
building '/nix/store/m9mg8cyy03r4lv7dfbnm0402sqrmchgp-odin-dev-2025-01.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/jxbz0g6x3ww8ab3ialnxprsrabli4gqm-source
source root is source
Running phase: patchPhase
applying patch /nix/store/csjm4s9f8i81jl63xc7qqff1zk3g4qxh-darwin-remove-impure-links.patch
patching file src/linker.cpp
Hunk #1 succeeded at 648 (offset -121 lines).
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: buildPhase
build flags: SHELL=/nix/store/xg75pc4yyfd5n2fimhb98ps910q5lm5n-bash-5.2p37/bin/bash release
./build_odin.sh release
/nix/store/xg75pc4yyfd5n2fimhb98ps910q5lm5n-bash-5.2p37/bin/bash: line 1: ./build_odin.sh: cannot execute: required file not found
make: *** [Makefile:16: release] Error 127
error: builder for '/nix/store/m9mg8cyy03r4lv7dfbnm0402sqrmchgp-odin-dev-2025-01.drv' failed with exit code 2;
       last 13 log lines:
       > Running phase: unpackPhase
       > unpacking source archive /nix/store/jxbz0g6x3ww8ab3ialnxprsrabli4gqm-source
       > source root is source
       > Running phase: patchPhase
       > applying patch /nix/store/csjm4s9f8i81jl63xc7qqff1zk3g4qxh-darwin-remove-impure-links.patch
       > patching file src/linker.cpp
       > Hunk #1 succeeded at 648 (offset -121 lines).
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: buildPhase
       > build flags: SHELL=/nix/store/xg75pc4yyfd5n2fimhb98ps910q5lm5n-bash-5.2p37/bin/bash release
       > ./build_odin.sh release
       > /nix/store/xg75pc4yyfd5n2fimhb98ps910q5lm5n-bash-5.2p37/bin/bash: line 1: ./build_odin.sh: cannot execute: required file not found
       > make: *** [Makefile:16: release] Error 127
       For full logs, run:
         nix log /nix/store/m9mg8cyy03r4lv7dfbnm0402sqrmchgp-odin-dev-2025-01.drv
```

### Additional context

I first ran into this when updating my system using `nix flake update` and `nixos-rebuild switch --flake <path/to/flake>#default`.
But it also happens by manually building with `nix-build` as shown in the reproducing steps.

I have tried fixing this by modifying the build phase in the package but nothing works.
For some reason it just refuses to find `build_odin.sh`.

Setting the `buildPhase` to this for debugging:
```odin
buildPhase = ''
  ls
  make
'';
```
shows the `build_odin.sh` in the `ls` output, but just refuses to find it when running `make`.
(running `./build_odin.sh` here instead of `make` it also does not find it)

log when using the modified `buildPhase`:
```log
$ nix-build -A odin
this derivation will be built:
  /nix/store/wk7pjh3kkhxajhgfwrhr9mvgihp3f7n6-odin-dev-2025-01.drv
building '/nix/store/wk7pjh3kkhxajhgfwrhr9mvgihp3f7n6-odin-dev-2025-01.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/jxbz0g6x3ww8ab3ialnxprsrabli4gqm-source
source root is source
Running phase: patchPhase
applying patch /nix/store/csjm4s9f8i81jl63xc7qqff1zk3g4qxh-darwin-remove-impure-links.patch
patching file src/linker.cpp
Hunk #1 succeeded at 648 (offset -121 lines).
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: buildPhase
LICENSE              README.md  build_odin.sh     examples  shell.nix
LLVM-C.dll           base       build_vendor.bat  misc      src
Makefile             bin        ci                odin.rdi  tests
PROPOSAL-PROCESS.md  build.bat  core              shared    vendor
PROGRAM=make ./build_odin.sh # debug
sh: line 1: ./build_odin.sh: cannot execute: required file not found
make: *** [Makefile:10: default] Error 127
error: builder for '/nix/store/wk7pjh3kkhxajhgfwrhr9mvgihp3f7n6-odin-dev-2025-01.drv' failed with exit code 2;
       last 16 log lines:
       > Running phase: unpackPhase
       > unpacking source archive /nix/store/jxbz0g6x3ww8ab3ialnxprsrabli4gqm-source
       > source root is source
       > Running phase: patchPhase
       > applying patch /nix/store/csjm4s9f8i81jl63xc7qqff1zk3g4qxh-darwin-remove-impure-links.patch
       > patching file src/linker.cpp
       > Hunk #1 succeeded at 648 (offset -121 lines).
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: buildPhase
       > LICENSE                  README.md  build_odin.sh     examples  shell.nix
       > LLVM-C.dll             base       build_vendor.bat  misc      src
       > Makefile             bin        ci                odin.rdi  tests
       > PROPOSAL-PROCESS.md  build.bat        core              shared    vendor
       > PROGRAM=make ./build_odin.sh # debug
       > sh: line 1: ./build_odin.sh: cannot execute: required file not found
       > make: *** [Makefile:10: default] Error 127
       For full logs, run:
         nix log /nix/store/wk7pjh3kkhxajhgfwrhr9mvgihp3f7n6-odin-dev-2025-01.drv
```

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.23, NixOS, 25.05 (Warbler), 25.05.20250417.b024ced`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.1`
 - channels(root): `"nixos-24.05"`
 - nixpkgs: `/nix/store/isfbldda5j8j6x3nbv1zim0c0dpf90v8-source`

### Notify maintainers

@Astavie 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/od/odin/package.nix
+++ b/pkgs/by-name/od/odin/package.nix
@@ -24,6 +24,9 @@ stdenv.mkDerivation {
   patches = [
     ./darwin-remove-impure-links.patch
   ];
+  postPatch = ''
+    patchShebangs --build build_odin.sh
+  '';
 
   LLVM_CONFIG = "${llvmPackages.llvm.dev}/bin/llvm-config";
 

